cpp/src/copying/gather.cu

```diff
@@ -74,17 +39,37 @@ std::unique_ptr<table> gather(table_view const& source_table,
 {
   CUDF_EXPECTS(gather_map.has_nulls() == false, "gather_map contains nulls");
 
-  std::unique_ptr<table> destination_table = cudf::type_dispatcher(gather_map.type(),
-                                                                   dispatch_map_type{},
-                                                                   source_table,
-                                                                   gather_map,
-                                                                   gather_map.size(),
-                                                                   bounds,
-                                                                   neg_indices,
-                                                                   mr,
-                                                                   stream);
+  // create index type normalizing iterator for the gather_map
+  auto map_begin = indexalator_factory::make_input_iterator(gather_map);
```

Comment:
Looking at the indexalator factory functions, there should be a check to require that the input column does not contain nulls.

Reply:
This makes sense. I have use cases for dictionary where I want to do operations on dictionary indices along with a null mask  generally using pair iterators. But to build a pair iterator requires sometimes allowing the indexalator to be created with a nullable column.

Reply:
> But to build a pair iterator requires sometimes allowing the indexalator to be created with a nullable column.

Sure, but this should use a private code path. The public factories should disallow columns with nulls.

Reply:
OK. Adding the exception will not effect this PR but will break some code unrelated to this PR so I would rather make this change in a separate PR. In that PR I would want to move the current code logic that creates a pair-indexalator to a utility or the indexalator factory too.
https://github.com/rapidsai/cudf/blob/48e9d1bb15c73f241a173bb6de02dbf496aa87bf/cpp/src/dictionary/replace.cu#L72-L77

Reply:
Created #6469 to track this